graylog2-server/src/main/java/org/graylog2/indexer/datanode/DatanodeMigrationLockServiceImpl.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.indexer.datanode;
+
+import com.github.rholder.retry.Attempt;
+import com.github.rholder.retry.RetryException;
+import com.github.rholder.retry.RetryListener;
+import com.github.rholder.retry.RetryerBuilder;
+import com.github.rholder.retry.StopStrategies;
+import com.github.rholder.retry.WaitStrategies;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import jakarta.inject.Inject;
+import org.graylog2.cluster.lock.Lock;
+import org.graylog2.cluster.lock.LockService;
+import org.graylog2.indexer.IndexSet;
+import org.graylog2.plugin.Tools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+public class DatanodeMigrationLockServiceImpl implements DatanodeMigrationLockService {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DatanodeMigrationLockServiceImpl.class);
+    public static final int LOCK_EXTEND_PERIOD_SECONDS = 10;
+    private static final String LOCK_RESOURCE_PREFIX = "remote-reindex-migration_";
+
+    private final LockService lockService;
+
+
+    private final Set<Lock> activeLocks = Collections.synchronizedSet(new HashSet<>());
+
+    @Inject
+    public DatanodeMigrationLockServiceImpl(LockService lockService) {
+        this.lockService = lockService;
+        startLocksExtendingThread(lockService);
+    }
+
+    private void startLocksExtendingThread(LockService lockService) {
+        ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor(new ThreadFactoryBuilder().setNameFormat("migration-locks-service-backend-%d").setDaemon(true).setUncaughtExceptionHandler(new Tools.LogUncaughtExceptionHandler(LOG)).build());
+        executorService.scheduleAtFixedRate(() -> {
+            final Set<Optional<Lock>> extendedLocks = activeLocks.stream().map(lockService::extendLock).collect(Collectors.toSet());
```

Comment:
1. Shouldn't this piece of code be synchronized as well?
2. Shouldn't `extendedLocks` replace the contents of `activeLocks` after the extending was completed?

Reply:
The updated_at field is not important from the datanode locking service code. Updating the lock inside the activeLocks collection would make it harder to remove it when releasing a lock, as the equals method compares also the updated_at field. And as the lock is updated in an async way, without propagating the updated lock to the original caller, I'd need an additional way to connect the original lock with the updated one. 

The only problem I can imagine is that we change in the future the implementation of the underlying locking service and it will reject extend requests with outdated locks. But as the api and functionality is designed now, I think we can safely ignore the return value of the extendLock call.